File: src/java/br/com/lett/crawlernode/crawlers/ranking/keywords/riodejaneiro/RiodejaneiroSuperprixCrawler.java
package br.com.lett.crawlernode.crawlers.ranking.keywords.riodejaneiro;

import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;
import br.com.lett.crawlernode.core.session.Session;
import br.com.lett.crawlernode.core.task.impl.CrawlerRankingKeywords;
import br.com.lett.crawlernode.util.CommonMethods;

public class RiodejaneiroSuperprixCrawler extends CrawlerRankingKeywords {

  public RiodejaneiroSuperprixCrawler(Session session) {
    super(session);
  }

  @Override
  protected void extractProductsFromCurrentPage() {
    this.log("Página " + this.currentPage);

    // número de produtos por página do market
    this.pageSize = 15;

    String keyword = this.keywordWithoutAccents.replaceAll(" ", "%20");

    // monta a url com a keyword e a página
    String url = "http://www.superprix.com.br/" + keyword + "?PageNumber=" + this.currentPage;
    this.log("Link onde são feitos os crawlers: " + url);

    // chama função de pegar a url
    this.currentDoc = fetchDocument(url);

    Elements products = this.currentDoc.select("div div.prateleira > ul > li[layout]");

    if (!products.isEmpty()) {
      if (this.totalProducts == 0) {
        setTotalProducts();
      }

      for (Element e : products) {
        String internalPid = crawlInternalPid(e);
        String productUrl = crawlProductUrl(e);

        saveDataProduct(null, internalPid, productUrl);

        this.log("Position: " + this.position + " - InternalId: " + null + " - InternalPid: " + internalPid + " - Url: " + productUrl);

        if (this.arrayProducts.size() == productsLimit) {
          break;
        }
      }
    } else {
      this.result = false;
      this.log("Keyword sem resultado!");
    }

    this.log("Finalizando Crawler de produtos da página " + this.currentPage + " - até agora " + this.arrayProducts.size() + " produtos crawleados");
  }

  @Override
  protected void setTotalProducts() {
    Element totalElement = this.currentDoc.select("span.resultado-busca-numero span.value").first();
    if (totalElement != null) {
      try {
        this.totalProducts = Integer.parseInt(totalElement.text());
      } catch (Exception e) {
        this.logError(CommonMethods.getStackTrace(e));
      }
    }

    this.log("Total da busca: " + this.totalProducts);
  }

  private String crawlInternalPid(Element e) {
    String internalPid = null;
    Element pidElement = e.select(".rating-produto[id]").first();

    if (pidElement != null) {
      internalPid = CommonMethods.getLast(pidElement.attr("id").split("-"));
    }

    return internalPid;
  }

  private String crawlProductUrl(Element e) {
    String urlProduct = null;
    Element urlElement = e.select(".data h3 a").first();

    if (urlElement != null) {
      urlProduct = urlElement.attr("href");
    }

    return urlProduct;
  }
}

File: src/java/br/com/lett/crawlernode/crawlers/ranking/keywords/riodejaneiro/RiodejaneiroSuperprixCrawler.kt
package br.com.lett.crawlernode.crawlers.ranking.keywords.riodejaneiro

import br.com.lett.crawlernode.core.models.RankingProductBuilder
import br.com.lett.crawlernode.core.session.Session
import br.com.lett.crawlernode.core.task.impl.CrawlerRankingKeywords
import br.com.lett.crawlernode.util.CrawlerUtils
import org.apache.http.impl.cookie.BasicClientCookie
import org.jsoup.nodes.Element
import java.util.*

class RiodejaneiroSuperprixCrawler(session: Session) : CrawlerRankingKeywords(session) {

   private val home = "https://www.superprix.com.br/"

   val token: String by lazy {
      fetchDocument(home).selectFirst("li[layout]").attr("layout")
   }

   init {
      pageSize = 15
   }

   override fun extractProductsFromCurrentPage() {
      val url = home + "buscapagina?ft=${keywordWithoutAccents.replace(" ", "+")}" +
         "&PS=15&sl=$token&cc=15&sm=0&PageNumber=$currentPage"
      val fetchDocument = fetchDocument(url, cookies)

      fetchDocument.select(".prateleira li[layout]")?.forEach {
         val internalId = it.selectFirst(".buy-button-normal")?.attr("id")
         val internalPid = it.selectFirst(".avaliacao div")?.attr("id")?.split("-")?.last()
         val productUrl = it.selectFirst("a")?.attr("href")
         val name = CrawlerUtils.scrapStringSimpleInfo(it, ".data > h3 > a", true);
         val imgUrl = CrawlerUtils.scrapSimplePrimaryImage(it, ".productImage img", Arrays.asList("src"), "https", "")
         val price = CrawlerUtils.scrapPriceInCentsFromHtml(it, ".newPrice em", null, false, ',', session, 0)

         val isAvailable = price != 0

         val productRanking = RankingProductBuilder.create()
            .setUrl(productUrl)
            .setInternalId(internalId)
            .setInternalPid(internalPid)
            .setName(name)
            .setImageUrl(imgUrl)
            .setPriceInCents(price)
            .setAvailability(isAvailable)
            .build()

         saveDataProduct(productRanking)
      }
   }

   override fun checkIfHasNextPage(): Boolean {
      return (arrayProducts.size % pageSize - currentPage) < 0
   }
}
